### PR TITLE
[Telemetry Operator] Custom filters & outputs enable unsupported mode

### PR DESCRIPTION
**Pull request by lindnerby**
**Description**

Changes proposed in this pull request:

- Replace flag allowing unsupported plugins with defined custom filter or output that makes the LogPipeline run in Unsupported Mode
- Add metric for number of unsupported plugins `telemetry_plugins_unsupported_total`
- Add status to condition
- Remove supported plugins list
- Refactor plugin validation
- Rename restart metric `telemetry_fluentbit_restarts_total`
- Support only single output
- Reject pipeline when output is not defined.

**Related issue(s)**
* https://github.com/kyma-project/kyma/issues/14556
* https://github.com/kyma-project/kyma/issues/14555

### CODE REVIEW

File: components/telemetry-operator/controllers/logpipeline_controller.go [hunk 1]
```diff
@@ -124,11 +121,7 @@ func (r *LogPipelineReconciler) Reconcile(ctx context.Context, req ctrl.Request)
...
 
-	if r.Syncer.EnableUnsupportedPlugin {
-		r.EnableUnsupportedPlugin.Set(1)
-	} else {
-		r.EnableUnsupportedPlugin.Set(0)
-	}
+	r.unsupportedTotal.Set(float64(r.Syncer.UnsupportedPluginsTotal))
 
 	if changed {
 		log.V(1).Info("Fluent bit configuration was updated. Restarting the daemon set")
```
Comment (NHingerl) [inline components/telemetry-operator/controllers/logpipeline_controller.go:127]:
```suggestion
		log.V(1).Info("Fluent Bit configuration was updated. Restarting the DaemonSet.")
```

File: components/telemetry-operator/internal/sync/sync.go [hunk 2]
```diff
@@ -73,7 +73,7 @@ func (s *LogPipelineSyncer) SyncAll(ctx context.Context, logPipeline *telemetryv
 		log.Error(err, "Failed to sync variables")
 		return false, err
 	}
-	err = s.syncEnableUnsupportedPluginMetric(ctx)
+	err = s.syncUnsupportedPluginsTotal(ctx)
 	if err != nil {
 		log.Error(err, "Failed to sync enable unsupported plugin metrics")
```
Comment (NHingerl) [inline components/telemetry-operator/internal/sync/sync.go:78]: Not sure what "sync enable" means?

File: components/telemetry-operator/internal/validation/plugin_validator.go [hunk 3]
```diff
@@ -4,102 +4,119 @@ import (
...
 
-	for _, outputPlugin := range logPipeline.Spec.Outputs {
-		if err := checkIfPluginIsValid("output", outputPlugin.Content, logPipeline, v.supportedOutputPlugins, v.deniedOutputPlugins, logPipelines); err != nil {
+// Validate returns an error if validation fails
+// because of using denied plugins or errors in match conditions
+// for filters or outputs.
+func (pv *pluginValidator) Validate(logPipeline *telemetryv1alpha1.LogPipeline, logPipelines *telemetryv1alpha1.LogPipelineList) error {
+	err := pv.validateFilters(logPipeline, logPipelines)
+	if err != nil {
+		return errors.Wrap(err, " error validating Filter plugins")
```
Comment (NHingerl) [inline components/telemetry-operator/internal/validation/plugin_validator.go:55]:
```suggestion
		return errors.Wrap(err, " error validating filter plugins")
```

I assume there's a reason why the sentence starts with a blank space and lower case :)

File: components/telemetry-operator/internal/validation/plugin_validator.go [hunk 4]
```diff
@@ -4,102 +4,119 @@ import (
...
+	}
+	err = pv.validateOutputs(logPipeline, logPipelines)
+	if err != nil {
+		return errors.Wrap(err, " error validating Output plugins")
```
Comment (NHingerl) [inline components/telemetry-operator/internal/validation/plugin_validator.go:59]:
```suggestion
		return errors.Wrap(err, " error validating output plugins")
```

File: components/telemetry-operator/internal/validation/plugin_validator.go [hunk 5]
```diff
@@ -4,102 +4,119 @@ import (
...
 	return nil
 }
 
-func checkIfPluginIsValid(pluginType, pluginContent string, logPipeline *telemetryv1alpha1.LogPipeline, supportedPlugins, deniedPlugins []string, logPipelines *telemetryv1alpha1.LogPipelineList) error {
-	var err error
-	var section map[string]string
-	var pluginName string
+func (pv *pluginValidator) validateOutputs(pipeline *telemetryv1alpha1.LogPipeline, pipelines *telemetryv1alpha1.LogPipelineList) error {
+	if len(pipeline.Spec.Output.Custom) == 0 {
+		return fmt.Errorf("no output is defined! One output needs to be defined")
```
Comment (NHingerl) [inline components/telemetry-operator/internal/validation/plugin_validator.go:75]: Why does this start with lower case? If it's lower case by purpose, keep the lower case.

```suggestion
		return fmt.Errorf("No output is defined. You must define one output!")
```

File: components/telemetry-operator/internal/validation/plugin_validator.go [hunk 6]
```diff
@@ -4,102 +4,119 @@ import (
...
-	return false
-}
-
-func getSectionName(section map[string]string) (string, error) {
-	if name, hasKey := section["name"]; hasKey {
+func getCustomName(custom map[string]string) (string, error) {
+	if name, hasKey := custom["name"]; hasKey {
 		return name, nil
 	}
 	return "", fmt.Errorf("configuration section does not have name attribute")
```
Comment (NHingerl) [inline components/telemetry-operator/internal/validation/plugin_validator.go:122]: Does this start with lower case by purpose?

File: components/telemetry-operator/internal/validation/plugin_validator_test.go [hunk 7]
```diff
@@ -11,287 +11,169 @@ import (
...
+	err := sut.Validate(logPipeline, logPipelines)
 
-	logPipelines := &telemetryv1alpha1.LogPipelineList{
-		Items: []telemetryv1alpha1.LogPipeline{*logPipeline},
-	}
-
-	err := pluginValidator.Validate(logPipeline, logPipelines)
-	require.NoError(t, err)
+	require.Error(t, err)
+	require.Equal(t, " error validating Output plugins: no output is defined! One output needs to be defined", err.Error())
```
Comment (NHingerl) [inline components/telemetry-operator/internal/validation/plugin_validator_test.go:21]:
```suggestion
	require.Equal(t, " error validating output plugins: No output is defined. You must define one output!", err.Error())
```

File: components/telemetry-operator/internal/validation/plugin_validator_test.go [hunk 8]
```diff
@@ -11,287 +11,169 @@ import (
...
 	logPipeline.ObjectMeta = metav1.ObjectMeta{Name: "foo"}
-
 	logPipelines := &telemetryv1alpha1.LogPipelineList{}
 
-	err := pluginValidator.Validate(logPipeline, logPipelines)
-	assert.Contains(t, err.Error(), "output plugin 'http' with match condition 'abc' is not allowed. Valid match conditions are: 'foo' (current logpipeline name)")
+	sut := NewPluginValidator([]string{}, []string{})
+	err := sut.Validate(logPipeline, logPipelines)
+
+	assert.Contains(t, err.Error(), " error validating Output plugins: plugin 'http' with match condition 'abc' is not allowed. Valid match conditions are: 'foo' (current logpipeline name)")
```
Comment (NHingerl) [inline components/telemetry-operator/internal/validation/plugin_validator_test.go:120]:
```suggestion
	assert.Contains(t, err.Error(), " error validating output plugins: Plugin 'http' with match condition 'abc' is not allowed. Valid match conditions are: 'foo' (current LogPipeline name)")
```

File: components/telemetry-operator/internal/validation/plugin_validator_test.go [hunk 9]
```diff
@@ -11,287 +11,169 @@ import (
...
 		},
 	}
 	logPipeline2.ObjectMeta = metav1.ObjectMeta{Name: "bar"}
 
-	err := pluginValidator.Validate(logPipeline2, logPipelines)
-	assert.Contains(t, err.Error(), "output plugin 'http' with match condition 'bar' is not allowed. Valid match conditions are: 'bar' (current logpipeline name) or '[foo]' (other existing logpipelines names)")
+	sut := NewPluginValidator([]string{}, []string{})
+	err := sut.Validate(logPipeline2, logPipelines)
+
+	assert.Contains(t, err.Error(), "plugin 'http' with match condition 'bar' is not allowed. Valid match conditions are: 'bar' (current logpipeline name) or '[foo]' (other existing logpipelines names)")
```
Comment (NHingerl) [inline components/telemetry-operator/internal/validation/plugin_validator_test.go:149]:
```suggestion
	assert.Contains(t, err.Error(), "plugin 'http' with match condition 'bar' is not allowed. Valid match conditions are: 'bar' (current LogPipeline name) or '[foo]' (other existing LogPipelines names)")
```

File: components/telemetry-operator/internal/validation/plugin_validator_test.go [hunk 10]
```diff
@@ -300,31 +182,73 @@ func TestEnableAllPlugins(t *testing.T) {
...
-	err := pluginValidator.Validate(logPipeline, logPipelines)
-	require.NoError(t, err)
-}
+	sut := NewPluginValidator([]string{"lua", "multiline"}, []string{})
+	err := sut.Validate(logPipeline, logPipelines)
 
-func TestDeniedPlugins(t *testing.T) {
-	pluginValidator := NewPluginValidator([]string{"lua", "multiline"}, []string{}, []string{"kubernetes"}, []string{})
+	require.Error(t, err)
+	assert.Contains(t, err.Error(), " error validating Filter plugins: plugin 'lua' is not supported. ")
```
Comment (NHingerl) [inline components/telemetry-operator/internal/validation/plugin_validator_test.go:190]:
```suggestion
	assert.Contains(t, err.Error(), " error validating filter plugins: Plugin 'lua' is not supported. ")
```

File: components/telemetry-operator/internal/validation/plugin_validator_test.go [hunk 11]
```diff
@@ -300,31 +182,73 @@ func TestEnableAllPlugins(t *testing.T) {
...
 	logPipeline.Name = "foo"
 	logPipelines := &telemetryv1alpha1.LogPipelineList{}
-	logPipeline.Spec.EnableUnsupportedPlugins = true
 
-	err := pluginValidator.Validate(logPipeline, logPipelines)
+	sut := NewPluginValidator([]string{}, []string{"lua", "multiline"})
+	err := sut.Validate(logPipeline, logPipelines)
+
 	require.Error(t, err)
+	assert.Contains(t, err.Error(), " error validating Output plugins: plugin 'lua' is not supported. ")
```
Comment (NHingerl) [inline components/telemetry-operator/internal/validation/plugin_validator_test.go:211]:
```suggestion
	assert.Contains(t, err.Error(), " error validating output plugins: Plugin 'lua' is not supported. ")
```

File: components/telemetry-operator/api/v1alpha1/logpipeline_types.go [hunk 12]
```diff
@@ -147,7 +147,6 @@ func filterOutCondition(conds []LogPipelineCondition, condType LogPipelineCondit
 //+kubebuilder:resource:scope=Cluster
 //+kubebuilder:subresource:status
 //+kubebuilder:printcolumn:name="Status",type=string,JSONPath=`.status.conditions[-1].type`
-//+kubebuilder:printcolumn:name="Unsupported-Plugins",type=boolean,JSONPath=`.spec.enableUnsupportedPlugins`
 //+kubebuilder:printcolumn:name="Age",type=date,JSONPath=`.metadata.creationTimestamp`
 
 // LogPipeline is the Schema for the logpipelines API
```
Comment (NHingerl) [inline components/telemetry-operator/api/v1alpha1/logpipeline_types.go:152]: If this was lower case by purpose, ignore my suggestion

```suggestion
// LogPipeline is the Schema for the LogPipelines API
```

File: components/telemetry-operator/controllers/logpipeline_controller.go [hunk 13]
```diff
@@ -44,35 +44,32 @@ var (
...
-	FluentBitRestartsCount  prometheus.Counter
-	EnableUnsupportedPlugin prometheus.Gauge
+	Scheme             *runtime.Scheme
+	Syncer             *sync.LogPipelineSyncer
+	FluentBitDaemonSet types.NamespacedName
+	restartsTotal      prometheus.Counter
+	unsupportedTotal   prometheus.Gauge
 }
 
 // NewLogPipelineReconciler returns a new LogPipelineReconciler using the given FluentBit config arguments
```
Comment (NHingerl) [inline components/telemetry-operator/controllers/logpipeline_controller.go:54]:
```suggestion
// NewLogPipelineReconciler returns a new LogPipelineReconciler using the given Fluent Bit config arguments
```

File: components/telemetry-operator/controllers/logpipeline_controller.go [hunk 14]
```diff
@@ -44,35 +44,32 @@ var (
...
-	)
+	lpr.Syncer = sync.NewLogPipelineSyncer(client, daemonSetConfig, emitterConfig)
 	lpr.FluentBitDaemonSet = daemonSetConfig.FluentBitDaemonSetName
 
-	lpr.FluentBitRestartsCount = prometheus.NewCounter(prometheus.CounterOpts{
-		Name: "telemetry_operator_fluentbit_restarts_total",
+	// Metrics
+	lpr.restartsTotal = prometheus.NewCounter(prometheus.CounterOpts{
+		Name: "telemetry_fluentbit_restarts_total",
 		Help: "Number of triggered FluentBit restarts",
```
Comment (NHingerl) [inline components/telemetry-operator/controllers/logpipeline_controller.go:65]:
```suggestion
		Help: "Number of triggered Fluent Bit restarts",
```

File: components/telemetry-operator/api/v1alpha1/logpipeline_types.go (hunk 12 above)
Comment (chrkl) [reply components/telemetry-operator/api/v1alpha1/logpipeline_types.go:152]: The lower-case logpipeline refers to the Kubernetes resources. It's correct in this case

File: components/telemetry-operator/api/v1alpha1/logpipeline_types.go [hunk 15]
```diff
@@ -147,7 +147,6 @@ func filterOutCondition(conds []LogPipelineCondition, condType LogPipelineCondit
 //+kubebuilder:resource:scope=Cluster
 //+kubebuilder:subresource:status
 //+kubebuilder:printcolumn:name="Status",type=string,JSONPath=`.status.conditions[-1].type`
-//+kubebuilder:printcolumn:name="Unsupported-Plugins",type=boolean,JSONPath=`.spec.enableUnsupportedPlugins`
```
Comment (chrkl) [inline components/telemetry-operator/api/v1alpha1/logpipeline_types.go:150]: Let's keep the print column and show the value from the new status field

File: components/telemetry-operator/api/v1alpha1/logpipeline_types.go [hunk 16]
```diff
@@ -47,12 +46,12 @@ type MultiLineParser struct {
 
 // Filter describes a Fluent Bit filter configuration section
```
Comment (chrkl) [inline components/telemetry-operator/api/v1alpha1/logpipeline_types.go:47]: Maybe we remove the "section" from the doc string. Having a raw config section in here will be the exception in future. Let's make it "Filter describes a Fluent Bit filter configuration". Same for output.

**Review by chrkl — CHANGES_REQUESTED** [review]

File: components/telemetry-operator/config/samples/telemetry_v1alpha1_logpipeline_loki.yaml [hunk 17]
```diff
@@ -4,31 +4,19 @@ metadata:
...
-        LineFormat         json
-        LogLevel           warn
-        LabelMapPath       /files/labelmap.json
+  multilineParsers: []
+  filters: []
+  output:
+    custom: |
+      Name               grafana-loki
+      Alias              loki-output
+      Match              kube.*
```
Comment (chrkl) [inline components/telemetry-operator/config/samples/telemetry_v1alpha1_logpipeline_loki.yaml:13]: Fix the Match expression

File: components/telemetry-operator/controllers/logpipeline_controller.go [hunk 18]
```diff
@@ -112,7 +109,8 @@ func (r *LogPipelineReconciler) Reconcile(ctx context.Context, req ctrl.Request)
 			telemetryv1alpha1.SecretsNotPresent,
 			telemetryv1alpha1.LogPipelinePending,
 		)
-		if err := r.updateLogPipelineStatus(ctx, req.NamespacedName, condition); err != nil {
+		pipleLineUnsupported := sync.LogPipelineIsUnsupported(&logPipeline)
```
Comment (chrkl) [inline components/telemetry-operator/controllers/logpipeline_controller.go:112]: Typo

File: components/telemetry-operator/controllers/logpipeline_controller.go [hunk 19]
```diff
@@ -124,14 +122,10 @@ func (r *LogPipelineReconciler) Reconcile(ctx context.Context, req ctrl.Request)
...
-	if r.Syncer.EnableUnsupportedPlugin {
-		r.EnableUnsupportedPlugin.Set(1)
-	} else {
-		r.EnableUnsupportedPlugin.Set(0)
-	}
+	r.unsupportedTotal.Set(float64(r.Syncer.UnsupportedPluginsTotal))
 
 	if changed {
-		log.V(1).Info("Fluent bit configuration was updated. Restarting the daemon set")
+		log.V(1).Info("Fluent Bit configuration was updated. Restarting the DaemonSet.")
```
Comment (chrkl) [inline components/telemetry-operator/controllers/logpipeline_controller.go:128]: Remove full stop from log message for consistency

File: components/telemetry-operator/controllers/logpipeline_controller.go [hunk 20]
```diff
@@ -147,7 +141,8 @@ func (r *LogPipelineReconciler) Reconcile(ctx context.Context, req ctrl.Request)
 			telemetryv1alpha1.FluentBitDSRestartedReason,
 			telemetryv1alpha1.LogPipelinePending,
 		)
-		if err = r.updateLogPipelineStatus(ctx, req.NamespacedName, condition); err != nil {
+		pipleLineUnsupported := sync.LogPipelineIsUnsupported(&logPipeline)
```
Comment (chrkl) [inline components/telemetry-operator/controllers/logpipeline_controller.go:144]: Typo

File: components/telemetry-operator/controllers/logpipeline_controller.go [hunk 21]
```diff
@@ -171,7 +166,9 @@ func (r *LogPipelineReconciler) Reconcile(ctx context.Context, req ctrl.Request)
 			telemetryv1alpha1.FluentBitDSRestartCompletedReason,
 			telemetryv1alpha1.LogPipelineRunning,
 		)
-		if err = r.updateLogPipelineStatus(ctx, req.NamespacedName, condition); err != nil {
+		pipleLineUnsupported := sync.LogPipelineIsUnsupported(&logPipeline)
```
Comment (chrkl) [inline components/telemetry-operator/controllers/logpipeline_controller.go:169]: Typo

File: components/telemetry-operator/internal/sync/sync.go (hunk 2 above)
Comment (chrkl) [inline components/telemetry-operator/internal/sync/sync.go:78]: Change plugin to mode also in the log message

File: components/telemetry-operator/internal/validation/plugin_validator_test.go (hunk 10 above)
Comment (rakesh-garimella) [reply components/telemetry-operator/internal/validation/plugin_validator_test.go]: fixed.

File: components/telemetry-operator/api/v1alpha1/logpipeline_types.go (hunk 12 above)
Comment (rakesh-garimella) [reply components/telemetry-operator/api/v1alpha1/logpipeline_types.go:153]: will leave it as it is

File: components/telemetry-operator/api/v1alpha1/logpipeline_types.go (hunk 15 above)
Comment (rakesh-garimella) [reply components/telemetry-operator/api/v1alpha1/logpipeline_types.go:150]: done.

File: components/telemetry-operator/internal/validation/plugin_validator_test.go (hunk 11 above)
Comment (rakesh-garimella) [reply components/telemetry-operator/internal/validation/plugin_validator_test.go]: fixed.

File: components/telemetry-operator/internal/sync/sync.go (hunk 2 above)
Comment (rakesh-garimella) [reply components/telemetry-operator/internal/sync/sync.go:78]: fixed.

**Review by chrkl — APPROVED** [review]

**Review by jeremyharisch — APPROVED** [review]

**Pull request merged**